doc/typing_syntax.md

```diff
@@ -0,0 +1,124 @@
+
+and **mappings** exist.
+
+| Docstring type       | Python type annotation |
+|----------------------|------------------------|
+| `MAPPING of {X: Y}`  | `MAPPING[X, Y]`        |
+| `dict of {str: int}` | `dict[str, int]`       |
+
+
+> [!TIP]
+> While it is possible to nest these variants repeatedly, it is discouraged to do so to keep type descriptions readable.
+> For complex annotations with nested containers, consider using Python's conventional syntax.
+> In the future, docstub may warn against or disallow nesting these natural language variants.
+
+
+## Shape and dtype syntax for arrays
+
+This expression allows adding shape and datatype information for data structures like [NumPy arrays](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html).
+
+`array` and `ndarray`, and `array-like` and `array_like` can be used interchange-ably.
+
+| Docstring type              | Python type annotation |
+|-----------------------------|------------------------|
+| `array of DTYPE`            | `ndarray[DTYPE]`       |
+| `ndarray of dtype DTYPE`    | `ndarray[DTYPE]`       |
+| `array-like of DTYPE`       | `ArrayLike[DTYPE]`     |
+| `array_like of dtype DTYPE` | `ArrayLike[DTYPE]`     |
+
+> [!NOTE]
+> Noting the **shape** of an array in the docstring is supported.
+> However, Python's typing system is not yet able to express this information.
+> It is therefore not included in the resulting type annotation.
+
+| Docstring type           | Python type annotation |
+|--------------------------|------------------------|
+| `(3,) array of DTYPE`    | `ndarray[DTYPE]`       |
+| `(X, Y) array of DTYPE`  | `ndarray[DTYPE]`       |
+| `([P,] M, N) array-like` | `ArrayLike`            |
+| `(M, ...) ndarray`       | `ArrayLike`            |
+
+
+## Literals
+
+[Literals](https://typing.python.org/en/latest/spec/literal.html#literals) indicate a concrete value instead of type.
+Instead of using [`typing.Literal`](https://docs.python.org/3/library/typing.html#typing.Literal), you can enclose literal values in `{...}` in docstrings.
+
+| Docstring type | Python type annotation |
+|----------------|------------------------|
+| `{1, 2, 3}`    | `Literal[1, 2, 3]`     |
+| `{1, 2, 3}`    | `Literal[1, 2, 3]`     |
+
+> [!TIP]
+> Enclosing a single value `{X}` is currently allowed but discouraged.
+> Instead consider the more explicit `Literal[X]`.
+
+
+## reStructuredText role
+
+Since docstrings are also used to generate documentation with Sphinx, you may want to use [restructuredText roles](https://docutils.sourceforge.io/docs/ref/rst/roles.html) in your type annotations.
+Docstub allows for this anywhere where a qualified name can be used.
+
+| Docstring type       | Python type annotation |
+|----------------------|------------------------|
+| `` `X` ``            | `X`                    |
+| ``:ref:`X` ``        | `X`                    |
+| ``:class:`Y.X` ``    | `Y.X`                  |
+| ``:py:class:`Y.X` `` | `Y.X`                  |
```

Comment:
This is very nice! I have actually used it already but didn't properly realize it was happening. I will test it out for potential edge cases (tildes, explicit text, ommited imports) and report back. Matplotlib has a lot of them for example: https://matplotlib.org/devdocs/devel/document.html#reference-types